harness/determined/cli/dev.py

```diff
@@ -61,6 +68,201 @@ def curl(args: Namespace) -> None:
+        from typing import get_args, get_origin  # type: ignore
+    except ImportError:
+        raise errors.CliError("python >= 3.8 is required to use this feature")
+
+    a = unwrap_optional(a)
+    if a in [str, int, float, type(None), bool]:
+        return True
+
+    origin = get_origin(a)
+    args = get_args(a)
+
+    if origin in [list, abc_Sequence]:
+        if args is not None:
+            return all(is_supported_annotation(arg) for arg in args)
+
+    return False
+
+
+def can_be_called_via_cli(params: List[inspect.Parameter]) -> bool:
+    """
+    if all the non-optionals are primitives, then we can call it via the cli
+    """
+    for p in params:
+        if p.default is not inspect.Parameter.empty:
+            continue
+        if not is_supported_annotation(p.annotation):
+            return False
+    return True
+
+
+def get_available_bindings(
+    show_unusable: bool = False,
+) -> OrderedDict[str, List[inspect.Parameter]]:
+    rv: List[Tuple[str, List[inspect.Parameter]]] = []
+    for name, obj in inspect.getmembers(bindings):
+        if not inspect.isfunction(obj):
+            continue
+        name, params = bindings_sig(obj)
+        if not show_unusable:
+            if not can_be_called_via_cli(params):
+                continue
+            params = [p for p in params if is_supported_annotation(p.annotation)]
+        rv.append((name, params))
+
+    rv.sort(key=lambda x: x[0])
+    odict: OrderedDict[str, List[inspect.Parameter]] = collections.OrderedDict()
+    for name, params in rv:
+        odict[name] = params
+    return odict
+
+
+def parse_param_value(param: inspect.Parameter, value: str) -> Any:
+    annot = unwrap_optional(param.annotation)
+
+    if annot is str:
+        return value
+    elif annot is float:
+        return float(value)
+    elif annot is int:
+        return int(value)
+    elif annot is bool:
+        assert value.lower() in ["true", "false"]
+        return value.lower() == "true"
+    try:
+        value = json.loads(value)
+    except json.JSONDecodeError:
+        raise ValueError(f"Invalid JSON for {param.name}")
+
+    return value
+
+
+def parse_args_to_kwargs(args: List[str], params: List[inspect.Parameter]) -> Dict[str, Any]:
+    kwargs: Dict[str, Any] = {}
+    params_d: Dict[str, inspect.Parameter] = {p.name: p for p in params}
+
+    for idx, arg in enumerate(args):
+        key: str = ""
+        value: Any = None
+        if "=" in arg:
+            key, value = arg.split("=", 1)
+        else:
+            key = params[idx].name
+            value = arg
+        param = params_d.get(key)
+        if not param:
+            raise ValueError(f"Unknown argument {key}")
+        value = parse_param_value(param, value)
+        if key in kwargs:
+            raise ValueError(f"Argument {key} specified twice")
+        kwargs[key] = value
+
+    assert len(kwargs) <= len(params), "too many arguments"
+    return kwargs
+
+
+def print_resposne(d: Any) -> None:
```

Comment:
I think it's just in general easier to test and reason about pure functions than about impure ones. With printing, I think it can also help maintain the boundary between view and controller. What do you think about changing the interface here to return a string that the controller prints?

Reply:
I agree but I don't intend to do more with this but if we want to go that route it might make sense to be housed under `cli.render` and then tested and changed to return a string instead.